File: src/rrr/rpc/heartbeat.cpp
module;

#include <stdint.h>
#include <stdlib.h>  // abort() referenced by rusty/function.hpp
#include <rusty/cell.hpp>
#include <rusty/function.hpp>
#include <rusty/rusty.hpp>
#include <time.h>

export module rrr.heartbeat;

import std;

export namespace rrr {

// Free helper backing the timespec→microseconds conversion used by
// `heartbeat_time_us`. Pure u64 arithmetic; the C++ wrapper owns the
// `clock_gettime` syscall and hands the resulting (sec, nsec) pair
// in. File-prefixed name avoids link-time collision with the
// equivalent helper in circuit_breaker.cpp. Authored as inline Rust
// DSL.
#if RUSTYCPP_RUST
fn heartbeat_timespec_to_us(tv_sec: u64, tv_nsec: u64) -> u64 {
    tv_sec * 1000000 + tv_nsec / 1000
}
#endif
/*RUSTYCPP:GEN-BEGIN id=heartbeat.1 version=1 rust_sha256=017382a17e9a7e3c5b8333c349a1793c7ab03dbd3d61e49fb09278fec3c88038*/
uint64_t heartbeat_timespec_to_us(uint64_t tv_sec, uint64_t tv_nsec);

uint64_t heartbeat_timespec_to_us(uint64_t tv_sec, uint64_t tv_nsec) {
    return (rusty::detail::deref_if_pointer_like(tv_sec) * static_cast<uint64_t>(1000000)) + (rusty::detail::deref_if_pointer_like(tv_nsec) / static_cast<uint64_t>(1000));
}
/*RUSTYCPP:GEN-END id=heartbeat.1*/

inline uint64_t heartbeat_time_us() {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return heartbeat_timespec_to_us(
        static_cast<uint64_t>(ts.tv_sec),
        static_cast<uint64_t>(ts.tv_nsec));
}

struct HeartbeatConfig {
    bool enabled;
    uint32_t interval_ms;
    uint32_t timeout_ms;
    uint32_t max_missed;

    HeartbeatConfig()
        : enabled(true)
        , interval_ms(10000)
        , timeout_ms(5000)
        , max_missed(3)
    {}

    HeartbeatConfig(
        bool enabled_,
        uint32_t interval_ms_,
        uint32_t timeout_ms_,
        uint32_t max_missed_
    )
        : enabled(enabled_)
        , interval_ms(interval_ms_)
        , timeout_ms(timeout_ms_)
        , max_missed(max_missed_)
    {}

    static HeartbeatConfig aggressive() {
        return HeartbeatConfig(true, 5000, 2000, 2);
    }

    static HeartbeatConfig relaxed() {
        return HeartbeatConfig(true, 30000, 15000, 5);
    }

    static HeartbeatConfig disabled() {
        return HeartbeatConfig(false, 0, 0, 0);
    }
};

// Free helpers backing the timing-arithmetic portions of
// `HeartbeatManager::should_send_heartbeat`, `check_timeout`, and
// `time_until_next_heartbeat_ms`. All three convert `_ms` to `_us`
// and compare against a "now - last" elapsed window. Authored as
// inline Rust DSL.
#if RUSTYCPP_RUST
fn heartbeat_interval_elapsed(now_us: u64, last_send_us: u64, interval_ms: u32) -> bool {
    let interval_us: u64 = (interval_ms as u64) * 1000;
    now_us - last_send_us >= interval_us
}

fn heartbeat_timeout_elapsed(now_us: u64, last_send_us: u64, timeout_ms: u32) -> bool {
    let timeout_us: u64 = (timeout_ms as u64) * 1000;
    now_us - last_send_us >= timeout_us
}

fn heartbeat_time_until_next_ms(now_us: u64, last_send_us: u64, interval_ms: u32) -> u32 {
    let interval_us: u64 = (interval_ms as u64) * 1000;
    let elapsed_us: u64 = now_us - last_send_us;
    if elapsed_us >= interval_us {
        return 0;
    }
    ((interval_us - elapsed_us) / 1000) as u32
}
#endif
/*RUSTYCPP:GEN-BEGIN id=heartbeat.2 version=1 rust_sha256=ad0dc63f92a1763ba1d8538d2f3dcce52c56109bf1c6a96eec09b5ea43d3e67d*/
bool heartbeat_interval_elapsed(uint64_t now_us, uint64_t last_send_us, uint32_t interval_ms);
bool heartbeat_timeout_elapsed(uint64_t now_us, uint64_t last_send_us, uint32_t timeout_ms);
uint32_t heartbeat_time_until_next_ms(uint64_t now_us, uint64_t last_send_us, uint32_t interval_ms);

bool heartbeat_interval_elapsed(uint64_t now_us, uint64_t last_send_us, uint32_t interval_ms) {
    const uint64_t interval_us = ((static_cast<uint64_t>(interval_ms))) * static_cast<uint64_t>(1000);
    return (rusty::detail::deref_if_pointer_like(now_us) - rusty::detail::deref_if_pointer_like(last_send_us)) >= rusty::detail::deref_if_pointer_like(interval_us);
}

bool heartbeat_timeout_elapsed(uint64_t now_us, uint64_t last_send_us, uint32_t timeout_ms) {
    const uint64_t timeout_us = ((static_cast<uint64_t>(timeout_ms))) * static_cast<uint64_t>(1000);
    return (rusty::detail::deref_if_pointer_like(now_us) - rusty::detail::deref_if_pointer_like(last_send_us)) >= rusty::detail::deref_if_pointer_like(timeout_us);
}

uint32_t heartbeat_time_until_next_ms(uint64_t now_us, uint64_t last_send_us, uint32_t interval_ms) {
    const uint64_t interval_us = ((static_cast<uint64_t>(interval_ms))) * static_cast<uint64_t>(1000);
    const uint64_t elapsed_us = rusty::detail::deref_if_pointer_like(now_us) - rusty::detail::deref_if_pointer_like(last_send_us);
    if (rusty::detail::deref_if_pointer_like(elapsed_us) >= rusty::detail::deref_if_pointer_like(interval_us)) {
        return static_cast<uint32_t>(0);
    }
    return static_cast<uint32_t>((((rusty::detail::deref_if_pointer_like(interval_us) - rusty::detail::deref_if_pointer_like(elapsed_us))) / 1000));
}
/*RUSTYCPP:GEN-END id=heartbeat.2*/

// @safe - Heartbeat tracker. Fields are rusty::Cell<T> for trivially-
// copyable interior mutability + rusty::Function<void()> for the timeout
// callback. No raw pointers, syscalls, or operator-overload chains.
class HeartbeatManager {
private:
    HeartbeatConfig config_;

    rusty::Cell<uint64_t> last_send_time_{0};
    rusty::Cell<uint64_t> last_recv_time_{0};

    rusty::Cell<uint32_t> missed_count_{0};
    rusty::Cell<bool> pending_pong_{false};
    rusty::Cell<bool> timed_out_{false};

    rusty::Function<void()> on_timeout_;

public:
    explicit HeartbeatManager(const HeartbeatConfig& config = HeartbeatConfig())
        : config_(config)
    {}

    void set_config(const HeartbeatConfig& config) {
        config_ = config;
        reset();
    }

    void set_on_timeout(rusty::Function<void()> callback) {
        on_timeout_ = std::move(callback);
    }

    bool should_send_heartbeat() const {
        if (!config_.enabled || timed_out_.get()) {
            return false;
        }

        if (pending_pong_.get()) {
            return false;
        }

        return heartbeat_interval_elapsed(
            heartbeat_time_us(),
            last_send_time_.get(),
            config_.interval_ms);
    }

    void on_heartbeat_sent() {
        if (!config_.enabled) return;

        last_send_time_.set(heartbeat_time_us());
        pending_pong_.set(true);
    }

    void on_pong_received() {
        if (!config_.enabled) return;

        last_recv_time_.set(heartbeat_time_us());
        pending_pong_.set(false);
        missed_count_.set(0);
        timed_out_.set(false);
    }

    bool check_timeout() {
        if (!config_.enabled || timed_out_.get()) {
            return false;
        }

        if (!pending_pong_.get()) {
            return false;
        }

        if (heartbeat_timeout_elapsed(
                heartbeat_time_us(),
                last_send_time_.get(),
                config_.timeout_ms)) {
            pending_pong_.set(false);
            uint32_t count = missed_count_.get() + 1;
            missed_count_.set(count);

            if (count >= config_.max_missed) {
                timed_out_.set(true);

                if (on_timeout_) {
                    on_timeout_();
                }
                return true;
            }
        }

        return false;
    }

    uint32_t time_until_next_heartbeat_ms() const {
        if (!config_.enabled || timed_out_.get() || pending_pong_.get()) {
            return config_.interval_ms;
        }

        return heartbeat_time_until_next_ms(
            heartbeat_time_us(),
            last_send_time_.get(),
            config_.interval_ms);
    }

    bool is_timed_out() const {
        return timed_out_.get();
    }

    uint32_t missed_count() const {
        return missed_count_.get();
    }

    bool is_pending_pong() const {
        return pending_pong_.get();
    }

    void reset() {
        last_send_time_.set(0);
        last_recv_time_.set(0);
        missed_count_.set(0);
        pending_pong_.set(false);
        timed_out_.set(false);
    }

    const HeartbeatConfig& config() const {
        return config_;
    }
};

} // export namespace rrr

File: crates/srpc/tests/heartbeat.rs
use srpc::rpc::heartbeat::{heartbeat_time_us, HeartbeatConfig, HeartbeatManager};
use std::cell::Cell;
use std::rc::Rc;

fn zero_config() -> HeartbeatConfig {
    HeartbeatConfig {
        enabled: false,
        interval_ms: 0,
        timeout_ms: 0,
        max_missed: 0,
    }
}

fn expire_pending(manager: &HeartbeatManager) {
    let timeout_us = (manager.config().timeout_ms as u64) * 1_000;
    manager.last_send_time.set(heartbeat_time_us() - timeout_us);
}

#[test]
fn presets_exactly_match_the_legacy_values() {
    let defaults = HeartbeatConfig::defaults();
    assert!(defaults.enabled);
    assert_eq!(defaults.interval_ms, 10_000);
    assert_eq!(defaults.timeout_ms, 5_000);
    assert_eq!(defaults.max_missed, 3);

    let via_new = HeartbeatConfig::new();
    assert_eq!(via_new.enabled, defaults.enabled);
    assert_eq!(via_new.interval_ms, defaults.interval_ms);
    assert_eq!(via_new.timeout_ms, defaults.timeout_ms);
    assert_eq!(via_new.max_missed, defaults.max_missed);

    let aggressive = HeartbeatConfig::aggressive();
    assert!(aggressive.enabled);
    assert_eq!(aggressive.interval_ms, 5_000);
    assert_eq!(aggressive.timeout_ms, 2_000);
    assert_eq!(aggressive.max_missed, 2);

    let relaxed = HeartbeatConfig::relaxed();
    assert!(relaxed.enabled);
    assert_eq!(relaxed.interval_ms, 30_000);
    assert_eq!(relaxed.timeout_ms, 15_000);
    assert_eq!(relaxed.max_missed, 5);

    let disabled = HeartbeatConfig::disabled();
    assert!(!disabled.enabled);
    assert_eq!(disabled.interval_ms, 0);
    assert_eq!(disabled.timeout_ms, 0);
    assert_eq!(disabled.max_missed, 0);
}

#[test]
fn zero_initialized_config_and_disabled_preset_are_inert() {
    let zeroed = HeartbeatManager::new(&zero_config());
    assert!(!zeroed.should_send_heartbeat());
    zeroed.on_heartbeat_sent();
    assert!(!zeroed.is_pending_pong());
    assert!(!zeroed.check_timeout());

    let disabled = HeartbeatManager::new(&HeartbeatConfig::disabled());
    assert!(!disabled.should_send_heartbeat());
    assert!(!disabled.is_timed_out());
    assert_eq!(disabled.missed_count(), 0);
}

#[test]
fn interval_and_pending_pong_gate_heartbeat_sends() {
    let mut config = HeartbeatConfig::defaults();
    config.interval_ms = 50;
    let manager = HeartbeatManager::new(&config);

    assert!(manager.should_send_heartbeat(), "the first ping is due");
    manager.on_heartbeat_sent();
    assert!(manager.is_pending_pong());
    assert!(!manager.should_send_heartbeat());
    assert_eq!(manager.time_until_next_heartbeat_ms(), 50);

    manager.on_pong_received();
    assert!(!manager.is_pending_pong());
    assert!(!manager.should_send_heartbeat());
    manager.last_send_time.set(heartbeat_time_us() - 50_000);
    assert!(manager.should_send_heartbeat());
    assert_eq!(manager.time_until_next_heartbeat_ms(), 0);
}

#[test]
fn misses_accumulate_and_timeout_callback_fires_once() {
    let mut config = HeartbeatConfig::defaults();
    config.timeout_ms = 10;
    config.max_missed = 2;
    let manager = HeartbeatManager::new(&config);
    let callback_count = Rc::new(Cell::new(0_u32));
    let sink = Rc::clone(&callback_count);
    manager.set_on_timeout(Box::new(move || sink.set(sink.get() + 1)));

    manager.on_heartbeat_sent();
    expire_pending(&manager);
    assert!(!manager.check_timeout());
    assert_eq!(manager.missed_count(), 1);
    assert!(!manager.is_pending_pong());

    manager.on_heartbeat_sent();
    expire_pending(&manager);
    assert!(manager.check_timeout());
    assert_eq!(manager.missed_count(), 2);
    assert!(manager.is_timed_out());
    assert_eq!(callback_count.get(), 1);

    assert!(!manager.check_timeout());
    assert_eq!(callback_count.get(), 1);
}

#[test]
fn pong_reset_and_config_replacement_clear_runtime_state() {
    let mut config = HeartbeatConfig::defaults();
    config.timeout_ms = 0;
    config.max_missed = 1;
    let manager = HeartbeatManager::new(&config);

    manager.on_heartbeat_sent();
    assert!(manager.check_timeout());
    assert!(manager.is_timed_out());
    manager.on_pong_received();
    assert!(!manager.is_timed_out());
    assert_eq!(manager.missed_count(), 0);
    assert!(!manager.is_pending_pong());
    assert_ne!(manager.last_recv_time.get(), 0);

    manager.on_heartbeat_sent();
    let replacement = HeartbeatConfig::relaxed();
    manager.set_config(&replacement);
    assert!(!manager.is_timed_out());
    assert_eq!(manager.missed_count(), 0);
    assert!(!manager.is_pending_pong());
    assert_eq!(manager.last_send_time.get(), 0);
    assert_eq!(manager.last_recv_time.get(), 0);
    assert_eq!(manager.config().interval_ms, 30_000);
}

#[test]
fn replacing_the_callback_discards_the_old_observer() {
    let mut config = HeartbeatConfig::defaults();
    config.timeout_ms = 0;
    config.max_missed = 1;
    let manager = HeartbeatManager::new(&config);
    let first = Rc::new(Cell::new(0_u32));
    let first_sink = Rc::clone(&first);
    manager.set_on_timeout(Box::new(move || first_sink.set(first_sink.get() + 1)));

    let second = Rc::new(Cell::new(0_u32));
    let second_sink = Rc::clone(&second);
    manager.set_on_timeout(Box::new(move || second_sink.set(second_sink.get() + 1)));

    manager.on_heartbeat_sent();
    assert!(manager.check_timeout());
    assert_eq!(first.get(), 0);
    assert_eq!(second.get(), 1);
}
